### ISSUE TITLE
[no-unused-vars][BUG] Not working with unused parameter as stated in latest update

### ISSUE THREAD

**Issue by NoPhaseNoKill** [bradzacher/eslint-plugin-typescript#288]
**Repro**

```JSON
{
      "files": ["src/**/*.ts","src/**/*.tsx"],
      "parser": "eslint-plugin-typescript/parser",
      "parserOptions": {
        "jsx": true,
        "useJSXTextNode": true,
      },
      "rules": {
        "no-unused-vars": "off",
        "typescript/no-unused-vars": ["error", { "vars": "all", "args": "after-used", "ignoreRestSiblings": false }]
      },
}
```

```TS

export type Renderable = JSX.Element[] | JSX.Element | string;

export type AdviserDetails = {
  name: string
}

generateRows = (adviserDetails: AdviserDetails) => {
    const rowFieldsForCard: [string, Renderable?][] = [
      ['Name', adviserDetails.name],
    ]

return rowFieldsForCard
      .filter(([label, value]) => !!value)
      .map(([label, value]) => {
        return (
          <Table.Row key={label}>
            <Table.Cell>{label}</Table.Cell>
            <Table.Cell>{value}</Table.Cell>
          </Table.Row>
        )
      })
}
```

**Expected Result**
no errors
**Actual Result**
'label' is defined but never used typescript/no-unused-vars (happens for filter function)
**Additional Info**

**Versions**

| package                    | version |
| -------------------------- | ------- |
| `eslint-plugin-typescript` | `^1.0.0-rc.2` |
| `typescript-eslint-parser` | `21.0.2` |
| `typescript`               | `3.1.1` |

edit: formatting

**Comment by j-f1** [comment]
Seems like this is behaving as expected, since the regular [ESLint demo](https://eslint.org/demo/#eyJ0ZXh0Ijoicm93RmllbGRzRm9yQ2FyZFxuICAgICAgLmZpbHRlcigoW2xhYmVsLCB2YWx1ZV0pID0+ICEhdmFsdWUpXG4gICAgICAubWFwKChbbGFiZWwsIHZhbHVlXSkgPT4ge1xuICAgICAgICByZXR1cm4gKFxuICAgICAgICAgIDxUYWJsZS5Sb3cga2V5PXtsYWJlbH0+XG4gICAgICAgICAgICA8VGFibGUuQ2VsbD57bGFiZWx9PC9UYWJsZS5DZWxsPlxuICAgICAgICAgICAgPFRhYmxlLkNlbGw+e3ZhbHVlfTwvVGFibGUuQ2VsbD5cbiAgICAgICAgICA8L1RhYmxlLlJvdz5cbiAgICAgICAgKVxuICAgICAgfSkiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjcsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7Im5vLXVudXNlZC12YXJzIjoyfSwiZW52Ijp7fX19) shows the same behavior. You can use this syntax instead:

```js
.filter(([, value]) => !!value)
```

**Comment by bradzacher** [comment]
I was literally just looking at the same thing!

It's an expected error. You can fix this via using the `args` option https://eslint.org/docs/rules/no-unused-vars#args

From their examples, this is valid code:
```
/*eslint no-unused-vars: ["error", {"args": "after-used"}]*/

(function(foo, bar, baz, qux) {
    return qux;
})();
```

**Comment by bradzacher** [comment]
note as well, typescript's unused checker works off of prefixing with `_`.
It's a convention that I've come to like, rather than just omiting it entirely:

```
/*eslint no-unused-vars: ["error", {"args": "after-used"}]*/

.filter(([_label, value]) => !!value)
```

**Comment by NoPhaseNoKill** [comment]
> Seems like this is behaving as expected, since the regular [ESLint demo](https://eslint.org/demo/#eyJ0ZXh0Ijoicm93RmllbGRzRm9yQ2FyZFxuICAgICAgLmZpbHRlcigoW2xhYmVsLCB2YWx1ZV0pID0+ICEhdmFsdWUpXG4gICAgICAubWFwKChbbGFiZWwsIHZhbHVlXSkgPT4ge1xuICAgICAgICByZXR1cm4gKFxuICAgICAgICAgIDxUYWJsZS5Sb3cga2V5PXtsYWJlbH0+XG4gICAgICAgICAgICA8VGFibGUuQ2VsbD57bGFiZWx9PC9UYWJsZS5DZWxsPlxuICAgICAgICAgICAgPFRhYmxlLkNlbGw+e3ZhbHVlfTwvVGFibGUuQ2VsbD5cbiAgICAgICAgICA8L1RhYmxlLlJvdz5cbiAgICAgICAgKVxuICAgICAgfSkiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjcsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7Im5vLXVudXNlZC12YXJzIjoyfSwiZW52Ijp7fX19) shows the same behavior. You can use this syntax instead:
> 
> ```js
> .filter(([, value]) => !!value)
> ```

The key difference here is that I'm using a tuple, and the behaviour completely changes if label is removed from the parameter (ie it's being used, just under the covers). Should there be another way I code this instead? Keeping in mind I'm fairly new to Typescript.



> I was literally just looking at the same thing!
> 
> It's an expected error. You can fix this via using the `args` option https://eslint.org/docs/rules/no-unused-vars#args
> 
> From their examples, this is valid code:
> 
> ```js
> /*eslint no-unused-vars: ["error", {"args": "after-used"}]*/
> 
> (function(foo, bar, baz, qux) {
>     return qux;
> })();
> ```

My config is as yours states, and I'm getting the exact same error. Have also tried with your _label and this still fails

**Comment by NoPhaseNoKill** [comment]
Ok cool, just tested it and it works.

This seems a little odd to be behaving this way. Let's say for instance someone comes back to refactor the code; and this parameter is now used. They have no context of what the parameter is (as it's unnamed) - and will have to unnecessarily follow through the code to figure out what's going on.

The _ syntax isn't working for me - I get same error. Is there another way I can keep the naming standard, without having to put eslint ignore comments specifically for these cases?

**Comment by armano2** [comment]
i'm using comments for that 
```ts
.filter(([/*label*/, value]) => !!value)
```

**Comment by bradzacher** [comment]
Oh man, I wasn't paying attention at all, sorry.
OFC the config I suggested won't work, because it's not a function arg, it's an array destruction.

The underscore notation only works with typescript's unused vars checks. Though you can make it work with eslint via argsIgnorePattern https://eslint.org/docs/rules/no-unused-vars#argsignorepattern

```js
/*eslint no-unused-vars: ["error", { "argsIgnorePattern": "^_" }]*/

.filter(([_label, value]) => !!value)
```

I'm going to go ahead and close this as it's working as intended
